demes/demes.py

```diff
@@ -1412,78 +1429,62 @@ def deme(
                     f"of existence for ancestor {ancestor} "
                     f"({anc.start_time}, {anc.end_time})"
                 )
-
-        # build the deme's epochs
-        if epochs is None:
-            # if epochs are not given, we assign a single epoch over that deme
-            if final_size is None:
-                final_size = initial_size
-            if end_time is None:
-                end_time = 0
-            if initial_size == final_size:
-                size_function = "constant"
-            else:
-                size_function = "exponential"
-            epochs = [
-                Epoch(
-                    start_time=start_time,
-                    end_time=end_time,
-                    initial_size=initial_size,
-                    final_size=final_size,
-                    size_function=size_function,
-                    selfing_rate=selfing_rate,
-                    cloning_rate=cloning_rate,
-                )
-            ]
+        # set default cloning and selfing rates
+        if "selfing_rate" in defaults:
+            selfing_rate = defaults["selfing_rate"]
+        elif "selfing_rate" in self.defaults:
+            selfing_rate = self.defaults["selfing_rate"]
         else:
-            # the last epoch needs to have an end time
-            if epochs[-1].end_time is None:
-                if end_time is not None:
-                    epochs[-1].end_time = end_time
+            selfing_rate = None
+        if "cloning_rate" in defaults:
+            cloning_rate = defaults["cloning_rate"]
+        elif "cloning_rate" in self.defaults:
+            cloning_rate = self.defaults["cloning_rate"]
+        else:
+            cloning_rate = None
+
+        # the last epoch's end time is set to deme's end time
+        if epochs[-1].end_time is None:
+            epochs[-1].end_time = end_time
+        # the first epoch's start time is set to deme's start time
+        if epochs[0].start_time is None:
+            epochs[0].start_time = start_time
+        elif epochs[0].start_time != start_time:
+            raise ValueError("deme and first epoch start times do not align")
+
+        # fill in all attributes of epochs
+        for i in range(len(epochs)):
+            # set the start and end times based on surrounding demes
+            if epochs[i].start_time is None:
+                epochs[i].start_time = epochs[i - 1].end_time
+            if epochs[i].end_time is None:
+                if epochs[i + 1].start_time is None:
+                    raise ValueError("ambiguity about epochs' start/end times")
```

Comment:
The exception here should mention the deme id, and maybe also the indexes `i`/`i+1`?